File: packages/core/src/shadow/__tests__/APIRegistry.test.ts
import { afterEach, describe, expect, it } from 'vitest';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';

import { APIRegistry, APIRegistryEntry } from '../APIRegistry.js';

describe('APIRegistry', () => {
  const tempDirs: string[] = [];

  afterEach(() => {
    for (const dir of tempDirs) {
      try {
        fs.rmSync(dir, { recursive: true, force: true });
      } catch {
        // ignore cleanup errors
      }
    }
    tempDirs.length = 0;
  });

  it('register should add entry when no duplicate exists', async () => {
    const registry = new APIRegistry();

    const entry: APIRegistryEntry = {
      path: '/api/v1/users',
      method: 'GET',
      description: 'List all users',
      handler: 'UserController.list',
      tags: ['user', 'list'],
    };

    const result = await registry.register(entry);
    expect(result.isDuplicate).toBe(false);
    expect(result.similarEntries).toHaveLength(0);
    expect(registry.getEntryCount()).toBe(1);
  });

  it('checkDuplicate should detect exact path+method duplicate', async () => {
    const registry = new APIRegistry();

    const entry1: APIRegistryEntry = {
      path: '/api/v1/users',
      method: 'GET',
      description: 'List all users',
      handler: 'UserController.list',
      tags: ['user', 'list'],
    };

    const entry2: APIRegistryEntry = {
      path: '/api/v1/users',
      method: 'GET',
      description: 'Get user list',
      handler: 'UserHandler.getAll',
      tags: ['user'],
    };

    await registry.register(entry1);
    const result = registry.checkDuplicate(entry2);
    expect(result.isDuplicate).toBe(true);
    expect(result.similarEntries.length).toBeGreaterThan(0);
    expect(result.similarEntries[0].similarity).toBe(1.0);
  });

  it('checkDuplicate should detect semantically similar APIs', async () => {
    const registry = new APIRegistry({ similarityThreshold: 0.4 });

    const entry1: APIRegistryEntry = {
      path: '/api/v1/users',
      method: 'GET',
      description: 'List all users with pagination',
      handler: 'UserController.list',
      tags: ['user', 'list', 'pagination'],
    };

    const entry2: APIRegistryEntry = {
      path: '/api/v2/users/list',
      method: 'GET',
      description: 'Get all users with pagination support',
      handler: 'UserHandler.listAll',
      tags: ['user', 'list', 'pagination'],
    };

    await registry.register(entry1);
    const result = registry.checkDuplicate(entry2);
    expect(result.isDuplicate).toBe(true);
    expect(result.similarEntries.length).toBeGreaterThan(0);
  });

  it('search should find matching APIs', async () => {
    const registry = new APIRegistry();

    await registry.register({
      path: '/api/v1/users',
      method: 'GET',
      description: 'List all users',
      handler: 'UserController.list',
      tags: ['user', 'list'],
    });

    await registry.register({
      path: '/api/v1/orders',
      method: 'POST',
      description: 'Create a new order',
      handler: 'OrderController.create',
      tags: ['order', 'create'],
    });

    await registry.register({
      path: '/api/v1/users/:id',
      method: 'GET',
      description: 'Get user by ID',
      handler: 'UserController.getById',
      tags: ['user', 'detail'],
    });

    const results = registry.search('user');
    expect(results.length).toBeGreaterThanOrEqual(2);
    expect(results.every((r) => r.similarity !== undefined && r.similarity > 0)).toBe(true);
  });

  it('search should return empty for unrelated query', async () => {
    const registry = new APIRegistry();

    await registry.register({
      path: '/api/v1/users',
      method: 'GET',
      description: 'List all users',
      handler: 'UserController.list',
      tags: ['user', 'list'],
    });

    const results = registry.search('zzzzzzzzz');
    expect(results).toHaveLength(0);
  });

  it('saveToYaml and loadFromYaml should persist entries', async () => {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'api-registry-'));
    tempDirs.push(tmpDir);

    const registryPath = path.join(tmpDir, '.codeflow', 'registry', 'apis.yaml');

    const registry1 = new APIRegistry({ registryPath });
    await registry1.register({
      path: '/api/v1/users',
      method: 'GET',
      description: 'List all users',
      handler: 'UserController.list',
      tags: ['user', 'list'],
    });
    await registry1.register({
      path: '/api/v1/orders',
      method: 'POST',
      description: 'Create order',
      handler: 'OrderController.create',
      tags: ['order'],
    });
    await registry1.saveToYaml();

    expect(fs.existsSync(registryPath)).toBe(true);

    const registry2 = new APIRegistry({ registryPath });
    await registry2.loadFromYaml();
    expect(registry2.getEntryCount()).toBe(2);

    const entries = registry2.getEntries();
    expect(entries[0].path).toBe('/api/v1/users');
    expect(entries[1].path).toBe('/api/v1/orders');
  });

  it('loadFromYaml should handle missing file gracefully', async () => {
    const registry = new APIRegistry({ registryPath: '/nonexistent/path/apis.yaml' });
    await registry.loadFromYaml();
    expect(registry.getEntryCount()).toBe(0);
  });
});

File: backend/internal/shadow/api_registry_test.go
package shadow

import (
	"os"
	"path/filepath"
	"testing"
)

func TestAPIRegistryRegister(t *testing.T) {
	registry := NewAPIRegistry(nil)

	entry := APIRegistryEntry{
		Path:        "/api/v1/users",
		Method:      "GET",
		Description: "List all users",
		Handler:     "UserController.list",
		Tags:        []string{"user", "list"},
	}

	result := registry.Register(entry)
	if result.IsDuplicate {
		t.Fatal("expected no duplicate on first register")
	}
	if registry.GetEntryCount() != 1 {
		t.Fatalf("expected 1 entry, got %d", registry.GetEntryCount())
	}
}

func TestAPIRegistryCheckDuplicateExact(t *testing.T) {
	registry := NewAPIRegistry(nil)

	entry1 := APIRegistryEntry{
		Path:        "/api/v1/users",
		Method:      "GET",
		Description: "List all users",
		Handler:     "UserController.list",
		Tags:        []string{"user", "list"},
	}
	registry.Register(entry1)

	entry2 := APIRegistryEntry{
		Path:        "/api/v1/users",
		Method:      "GET",
		Description: "Get user list",
		Handler:     "UserHandler.getAll",
		Tags:        []string{"user"},
	}

	result := registry.CheckDuplicate(entry2)
	if !result.IsDuplicate {
		t.Fatal("expected duplicate for same path+method")
	}
	if len(result.SimilarEntries) == 0 {
		t.Fatal("expected at least one similar entry")
	}
	if result.SimilarEntries[0].Similarity != 1.0 {
		t.Fatalf("expected similarity 1.0, got %f", result.SimilarEntries[0].Similarity)
	}
}

func TestAPIRegistryCheckDuplicateSemantic(t *testing.T) {
	registry := NewAPIRegistry(&APIRegistryConfig{SimilarityThreshold: 0.4})

	entry1 := APIRegistryEntry{
		Path:        "/api/v1/users",
		Method:      "GET",
		Description: "List all users with pagination",
		Handler:     "UserController.list",
		Tags:        []string{"user", "list", "pagination"},
	}
	registry.Register(entry1)

	entry2 := APIRegistryEntry{
		Path:        "/api/v2/users/list",
		Method:      "GET",
		Description: "Get all users with pagination support",
		Handler:     "UserHandler.listAll",
		Tags:        []string{"user", "list", "pagination"},
	}

	result := registry.CheckDuplicate(entry2)
	if !result.IsDuplicate {
		t.Fatal("expected semantic duplicate")
	}
}

func TestAPIRegistrySearch(t *testing.T) {
	registry := NewAPIRegistry(nil)

	registry.Register(APIRegistryEntry{
		Path:        "/api/v1/users",
		Method:      "GET",
		Description: "List all users",
		Handler:     "UserController.list",
		Tags:        []string{"user", "list"},
	})
	registry.Register(APIRegistryEntry{
		Path:        "/api/v1/orders",
		Method:      "POST",
		Description: "Create a new order",
		Handler:     "OrderController.create",
		Tags:        []string{"order", "create"},
	})
	registry.Register(APIRegistryEntry{
		Path:        "/api/v1/users/:id",
		Method:      "GET",
		Description: "Get user by ID",
		Handler:     "UserController.getById",
		Tags:        []string{"user", "detail"},
	})

	results := registry.Search("user")
	if len(results) < 2 {
		t.Fatalf("expected at least 2 results, got %d", len(results))
	}
}

func TestAPIRegistrySearchNoMatch(t *testing.T) {
	registry := NewAPIRegistry(nil)

	registry.Register(APIRegistryEntry{
		Path:        "/api/v1/users",
		Method:      "GET",
		Description: "List all users",
		Handler:     "UserController.list",
		Tags:        []string{"user", "list"},
	})

	results := registry.Search("zzzzzzzzz")
	if len(results) != 0 {
		t.Fatalf("expected 0 results, got %d", len(results))
	}
}

func TestAPIRegistrySaveAndLoad(t *testing.T) {
	tmpDir, err := os.MkdirTemp("", "api_registry_test")
	if err != nil {
		t.Fatalf("create temp dir: %v", err)
	}
	defer os.RemoveAll(tmpDir)

	registryPath := filepath.Join(tmpDir, ".codeflow", "registry", "apis.yaml")

	r1 := NewAPIRegistry(&APIRegistryConfig{RegistryPath: registryPath})
	r1.Register(APIRegistryEntry{
		Path:        "/api/v1/users",
		Method:      "GET",
		Description: "List all users",
		Handler:     "UserController.list",
		Tags:        []string{"user", "list"},
	})
	r1.Register(APIRegistryEntry{
		Path:        "/api/v1/orders",
		Method:      "POST",
		Description: "Create order",
		Handler:     "OrderController.create",
		Tags:        []string{"order"},
	})

	if err := r1.SaveToYAML(); err != nil {
		t.Fatalf("save failed: %v", err)
	}

	r2 := NewAPIRegistry(&APIRegistryConfig{RegistryPath: registryPath})
	if err := r2.LoadFromYAML(); err != nil {
		t.Fatalf("load failed: %v", err)
	}

	if r2.GetEntryCount() != 2 {
		t.Fatalf("expected 2 entries, got %d", r2.GetEntryCount())
	}

	entries := r2.GetEntries()
	if entries[0].Path != "/api/v1/users" {
		t.Fatalf("unexpected first entry path: %s", entries[0].Path)
	}
}

func TestAPIRegistryLoadMissingFile(t *testing.T) {
	r := NewAPIRegistry(&APIRegistryConfig{RegistryPath: "/nonexistent/path/apis.yaml"})
	if err := r.LoadFromYAML(); err != nil {
		t.Fatalf("expected no error for missing file, got: %v", err)
	}
	if r.GetEntryCount() != 0 {
		t.Fatalf("expected 0 entries, got %d", r.GetEntryCount())
	}
}
